Basilisk master bc2 bjs2 1465

TrueSamedi opened the pull request.

TrueSamedi committed:
add RecommendationRequestDto
add RecommendationRequestController
add RecommendationRequestService (not full)

TrueSamedi committed:
add RecommendationRequestMapper
add SkillRequestDto
fixed RecommendationRequestDto
fixed RecommendationRequestService

TrueSamedi committed:
add RecommendationRequestServiceTest
add RecommendationRequestControllerTest
fixed RecommendationRequestDto
fixed RecommendationRequestService
fixed RecommendationRequestM
fixed RecommendationRequestService

TrueSamedi committed:
delete SkillRequestDto
add MessageRequestException
add RequestTimeOutException
add SkillsNotFoundException
add UserNotFoundException
fixed RecommendationRequestDto
fixed RecommendationRequestService
fixed RecommendationRequestMapper
fixed RecommendationRequestController
fixed RecommendationRequestControllerTest
fixed RecommendationRequestServiceTest

Rik-102 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,21 @@
+    public RecommendationRequestDto requestRecommendation(RecommendationRequestDto recommendationRequest) {
+        if (!(recommendationRequest.getMessage() != null) || !recommendationRequest.getMessage().isBlank() || !recommendationRequest.getMessage().isEmpty())
+            throw new MessageRequestException("Incorrect user's message");
+        return recommendationRequestService.create(recommendationRequest);
```
Давай запишем проще, без оператора "!". 
if (recommendationRequest.getMessage() == null) || recommendationRequest.getMessage().isBlank() || recommendationRequest.getMessage().isEmpty())
так легче воспринимается и читается)) 

и не хватает фигурных скобок после if(){
 тело метода;
};

знаю что можно в else их отпускать, но говорят это не является хорошей практикой)

Rik-102 commented on src/main/java/school/faang/user_service/dto/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.time.LocalDateTime;
+import java.util.List;
+
+@Data
+@AllArgsConstructor
```
Думаю нужно добавить @Builder, т.к.
очень удобно при тестах и просто удобно даже при созданий объектов)

Я еще @NoArgsConstructor но насколько это правильно, узнаю после ревью) поэтому за него ничего не говорю.

Rik-102 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,56 @@
+        Long receiverId = recommendationRequest.getReceiverId();
+        String message = recommendationRequest.getMessage();
+        Optional<RecommendationRequest> request = recommendationRequestRepository.findLatestPendingRequest(requesterId, receiverId);
+        Optional<User> requester = userRepository.findById(requesterId);
+        Optional<User> receiver = userRepository.findById(receiverId);
+
+        if (!requester.isPresent() && !receiver.isPresent())
+            throw new UserNotFoundException("User not found");
```
в целом работает, но можно заменить на orElseThrow(). будет легче читать и компактнее)

User requester = userRepository.findById(requesterId)
                                       .orElseThrow(()-> throw new UserNotFoundException("User not found"));
                                       
и для receiver ) 

Т.е. ты говоришь верни мне "requesterId", если нет такого id кинь эксепшн)

TrueSamedi replied:
и эта проверка не такая сейчас)) он в таком виде будет NullPointerException кидать вместо кастомного))
ша пойду разберусь что за магия произошла))

Rik-102 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,56 @@
+    private final RecommendationRequestMapper recommendationRequestMapper;
+
+    public RecommendationRequestDto create(RecommendationRequestDto recommendationRequest) {
+        Long requesterId = recommendationRequest.getRequesterId();
+        Long receiverId = recommendationRequest.getReceiverId();
+        String message = recommendationRequest.getMessage();
```
Это не нужно.  Есть маппер) 
если в ручную то на 53 строчке можно через get заполнить)

Rik-102 commented on src/main/java/school/faang/user_service/mapper/RecommendationRequestMapper.java:
```diff
@@ -0,0 +1,17 @@
+public interface RecommendationRequestMapper {
+
+    @Mapping(target = "message", source = "message")
+    @Mapping(target = "requesterId", source = "requesterId")
+    @Mapping(target = "receiverId", source = "receiverId")
```
не нужны
тк поля одинаковые

TrueSamedi replied:
реально магия. это не последний комит. я чет не то запушил

Rik-102 commented on src/test/java/school/faang/user_service/service/service/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,108 @@
+import org.mockito.Spy;
+import org.mockito.junit.jupiter.MockitoExtension;
+import school.faang.user_service.dto.RecommendationRequestDto;
+import school.faang.user_service.dto.SkillRequestDto;
```
Нет такого Dto -  SkillRequestDto;
и тесты не срабатывают

TrueSamedi replied:
странно. магия какая-то. в идее не подсвечивается как недобавленный. ша попробую перепушить

Rik-102 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,56 @@
+            }
+        }
+
+        recommendationRequest.getSkillsId()
+                .forEach(skillRequestId -> {
+                    if (!skillRequestRepository.existsById(skillRequestId)) {
+                        throw new SkillsNotFoundException("Skills not found");
+                    }
+                });
+        recommendationRequest.getSkillsId()
+                .forEach(skillRequestId -> skillRequestRepository.create(recommendationRequest.getId(), skillRequestId));
+
```
тут надо переделать, тк ты два раза итерируешься по  одному и тому же списку. 
можно через stream пропустить

TrueSamedi replied:
тут они должны логически быть разделены. нет смысла записывать скиллы, если хотя бы один не пройдет валидацию. создаст лишние сущности, которые нужно подтирать при эксепшене.

TrueSamedi committed:
add SkillRequestDto
add SkillRequestMapper
fixed RecommendationRequestDto
fixed RecommendationRequestMapper
fixed RecommendationRequestController
fixed RecommendationRequestService
fixed RecommendationRequestControllerTest
fixed RecommendationRequestServiceTest

TrueSamedi committed:
add SkillRequestDto
add SkillRequestMapper
fixed RecommendationRequestDto
fixed RecommendationRequestMapper
fixed RecommendationRequestController
fixed RecommendationRequestService
fixed RecommendationRequestControllerTest
fixed RecommendationRequestServiceTest

Rik-102 commented on src/main/java/school/faang/user_service/dto/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,22 @@
+    private Long id;
+    private String message;
+    private String status;
+    private List<SkillRequestDto> skillsId;
```
Возможно ты прав, что Дто надо передавать!

Напишу как бы сделал я!)

Сделал бы Лонг, так как по сути достаем просто число из сущности, и оно никак не ссылается сущность в БД. А там уже если нужно будет пользователю, то пусть делает отдельный запрос на сущность по тому числу - id которое он получил и тогда уже можно Дто отправить.
 
И тогда когда был бы тип Лонг.  В сервисе сделал бы  проверку через stream на skillId, чтобы не писать try- catch))

Но это лично моё мнение))

TrueSamedi replied:
я изначально тоже через id хотел, да и сейчас не до конца понимаю что мы еще там делать будем. но Паша сказал, что дальше вроде как дтошка понадобится)

Rik-102 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,50 @@
+
+    private final RecommendationRequestRepository recommendationRequestRepository;
+    private final SkillRequestRepository skillRequestRepository;
+    private final UserRepository userRepository;
```
вроде теперь юзер репо не задействован)

TrueSamedi replied:
о, не заметил. спасибо)

TrueSamedi committed:
fixed RecommendationRequestService

PavelAP96 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,21 @@
+@RequiredArgsConstructor
+public class RecommendationRequestController {
+
+    private RecommendationRequestService recommendationRequestService;
```
нет ключевого слова `final`
без него внедрение зависимостей при помощи аннотации `@RequiredArgsConstructor` работать не будет

PavelAP96 commented on src/main/java/school/faang/user_service/dto/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,22 @@
+    private Long id;
+    private String message;
+    private String status;
+    private List<SkillRequestDto> skillsId;
```
ты хранишь в этой переменной не айдишники скиллов, а скиллы, лучше назвать `skills`

PavelAP96 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,21 @@
+    private RecommendationRequestService recommendationRequestService;
+
+    public RecommendationRequestDto requestRecommendation(RecommendationRequestDto recommendationRequest) {
+        if ((recommendationRequest.getMessage() == null) || recommendationRequest.getMessage().isBlank() || recommendationRequest.getMessage().isEmpty())
```
`recommendationRequest.getMessage().isEmpty()` избыточно
проверка `isBlank` делает практически тоже самое, что и isEmpty, только еще проверяет, что строка не является строкой, которая содержит только пробелы

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,46 @@
+
+    public RecommendationRequestDto create(RecommendationRequestDto recommendationRequest) {
+
+        if (recommendationRequest.getRequesterId() == null || recommendationRequest.getReceiverId() == null)
+            throw new UserNotFoundException("User not found");
```
здесь ты не проверяешь, что requester и reciever присутствуют в базе данных
ты просто проверяешь, что у тебя в запросе значения не null
необходимо сделать проверку на наличие их в бд

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,46 @@
+        if (recommendationRequest.getRequesterId() == null || recommendationRequest.getReceiverId() == null)
+            throw new UserNotFoundException("User not found");
+        if (recommendationRequest.getCreatedAt().plusMonths(6).isAfter(recommendationRequest.getUpdatedAt())) {
+            throw new RequestTimeOutException("Last request was less than 6 months ago");
```
на самом деле не обязательно постоянно создавать новые исключения на каждый кейс, который ты обрабатываешь
уже создано достаточно много исключений, которые можно переиспользовать
например, в данном случае тебе можно использовать IllegalArgumentException с тем же сообщением, так как у тебя действительно на вход поданы данные, с которыми ты не можешь успешно завершить процесс
плодить все новые и новые исключения не нужно
только если действительно нет уже созданного исключения, которое бы тебе подходило

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,46 @@
+            throw new RequestTimeOutException("Last request was less than 6 months ago");
+        }
+
+        try {
+            recommendationRequest.getSkillsId()
+                    .forEach(skillRequestId -> {
+                        skillRequestRepository.existsById(skillRequestId.getSkillId());
+                    });
+        } catch (NullPointerException e) {
+            throw new SkillsNotFoundException("Skills not found");
+        }
```
твоя конструкция с try - catch не принесет желаемого результата
у тебя тут `NullPointerException` может выскочить только, если в `recommendationRequest` skillsId будет null, но это не значит, что у тебя скиллы не найдены в бд, а то, что у тебя запрос не валидный и скиллы не были переданы

я бы на твоем месте сделал отдельный метод, который будет у тебя валидировать запрос RecommendationRequestDto recommendationRequest и проверять, что все нужные поля у тебя не пустые

будешь вызывать его в основном методе в самом начале, а дальше прописывать оставшуюся логику, не перживая, что какое - то из полей пустое

PavelAP96 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,46 @@
+        recommendationRequest.getSkillsId()
+                .forEach(skillRequestId -> skillRequestRepository.create(skillRequestId.getId(), skillRequestId.getSkillId()));
+
+        return recommendationRequestMapper
```
чет ты все в одно налипил, выглядит не очень и тяжело читать
лучше сначала сохрани запрос в базу, сохрани в переменную
а потом вызови return с маппером, в который передашь эту переменную
